Name: <redacted> <redacted>

Github URL: https://github.com/<redacted>/assignment-5-judging-flowers-<redacted>


Commit ID: 6e0ccfd268db932c9780484c13066a91a80af6e3

In [9]:
# Imports and pip installations (if needed)
import pandas as pd 
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score

# Part 1: Load the dataset

In [10]:
# Load the dataset (load remotely or locally)
iris_df = pd.read_csv("/workspaces/assignment-5-judging-flowers-<redacted>/iris.csv")
# Output the first 15 rows of the data
print(iris_df.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(iris_df.info())

    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8         

In [11]:
print(iris_df.describe())

       sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000


## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?


This dataset is a collection of measurements taken from different types of iris flowers. For each flower, we have details about the lengths and widths of its sepals and petals. Using these measurements, we aim to determine the flower’s species, which could be one of three types: Iris setosa, Iris versicolor, or Iris virginica.



Features: The columns sepal_length, sepal_width, petal_length, and petal_width represent the measurements of the iris flowers. These are numerical values that describe the flower's physical dimensions.



Label: The species column represents the type of iris flower and serves as the label we’re trying to predict. There are three possible classes for this label: setosa, versicolor, and virginica.



Mapping of Labels to Classes


Each species name represents a unique class:
- setosa is one class label.
- versicolor is another.
- virginica is the third class.


In a classification task, models learn from the feature values to predict which class label (setosa, versicolor, or virginica) each instance of the data belongs to. 

# Part 2: Split the dataset into train and test

In [ ]:
# Take the dataset and split it into our features (X) and label (y)
X = iris_df.drop("species", axis=1)
y =  iris_df["species"]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size = 0.9, test_size=0.1, random_state=25)
# includig the shape of test and training set for the reference.
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

print("Number of samples in X_train:", len(X_train))
print("Number of samples in X_test:", len(X_test))
print("Number of samples in y_train:", len(y_train))
print("Number of samples in y_test:", len(y_test))

X_train shape: (135, 4)
X_test shape: (15, 4)
y_train shape: (135,)
y_test shape: (15,)
Number of samples in X_train: 135
Number of samples in X_test: 15
Number of samples in y_train: 135
Number of samples in y_test: 15


There are total 150 entries in the dataset.



90% of 150 is equal to 135 (that is the number of entries in training set).



10% of 150 is equal to 15 (that is the number of entries in test set).

# Part 3: Logistic Regression

In [ ]:
# i. Use sklearn to train a LogisticRegression model on the training set
model = LogisticRegression()
model.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
sample_dataPoint = pd.DataFrame([[6.5, 2.8, 4.6, 1.5]], columns=["sepal_length", "sepal_width", "petal_length", "petal_width"])
probs = model.predict_proba(sample_dataPoint)
for class_name, prob in zip(model.classes_, probs[0]):
    print(class_name, ":", round(prob, 4))
# iii. Report on the score for Logistic regression model, what does the score measure?
score = model.score(X_test, y_test)
print("score: ", score)
# generating the cross-validation score.
scores = cross_val_score(model, X, y, cv=8)
print("cross_val_score",scores)
print("Mean of cross-validation score:", scores.mean())
# iv. Extract the coefficents and intercepts for the boundary line(s)
coefficients = model.coef_
intercept = model.intercept_
print("Coefficients: ", coefficients)
print("Intercept: ", intercept)


setosa : 0.0027
versicolor : 0.8075
virginica : 0.1898
score:  0.9333333333333333
cross_val_score [1.         0.94736842 1.         0.89473684 0.94736842 0.94736842
 1.         1.        ]
Mean of cross-validation score: 0.9671052631578947
Coefficients:  [[-0.43966307  0.89157984 -2.43360878 -1.06690768]
 [ 0.56471311 -0.35214139 -0.18499237 -0.98045654]
 [-0.12505004 -0.53943845  2.61860115  2.04736421]]
Intercept:  [  9.84767041   2.01538136 -11.86305177]


The score tells us how often the model correctly predicts the class labels when tested on new data. A score of 0.9333 means that the model made the correct predictions 93.33% of the time on the test set. This high accuracy indicates that the model is performing well and is likely to make accurate predictions on similar data in the future.

Moreover the logistic regression model achieved a cross-validation mean score of 0.967. This score, which is close to 1, indicates the model's accuracy in correctly classifying the data points across different folds of the dataset. This high score implies that the model has a strong ability to generalize and is likely to perform well when applied to new, similar data.



Since we have 3 different classes in the dataset we can seperate the Coefficients and Intercept accordingly:

Class 0: 
- Coefficients: [-0.43966307  0.89157984 -2.43360878 -1.06690768]
- Intercept: 9.84767041

Class 1:
- Coefficients: [ 0.56471311 -0.35214139 -0.18499237 -0.98045654]
- Intercept: 2.01538136


Class 2:
- Coefficients: [-0.12505004 -0.53943845  2.61860115  2.04736421]
- Intercept: -11.86305177

Therefore, the equation for Class 0 is:

$Z = 9.84767041 - 0.43966307 \cdot sepalLength + 0.89157984 \cdot sepalWidth - 2.43360878 \cdot petalLength - 1.06690768 \cdot petalWidth$


For Class 1 is:

$Z = 2.01538136 + 0.56471311 \cdot sepalLength - 0.35214139 \cdot sepalWidth - 0.18499237 \cdot petalLength - 0.98045654 \cdot petalWidth$


And for Class 2 is:


$Z = -11.86305177 - 0.12505004 \cdot sepalLength - 0.53943845 \cdot sepalWidth + 2.61860115 \cdot petalLength + 2.04736421 \cdot petalWidth$



# Part 4: Support Vector Machine

In [ ]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svc_model = SVC(probability=True)  
svc_model.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
sample_dataPoint = pd.DataFrame([[5, 3.4, 1.5, 0.2]], columns=["sepal_length", "sepal_width", "petal_length", "petal_width"])
probs = svc_model.predict_proba(sample_dataPoint)
for class_name, prob in zip(svc_model.classes_, probs[0]):
    print(class_name, ":", round(prob, 4))
# iii. Report on the score for the SVM, what does the score measure?
svc_score = svc_model.score(X_test, y_test)
print("svc_score: ", svc_score)

# generating the cross-validation score.
scores = cross_val_score(svc_model, X, y, cv=5)
print("Cross-validation scores:", scores)
print("Mean of cross-validation score:", scores.mean())

setosa : 0.9693
versicolor : 0.0208
virginica : 0.0099
svc_score:  1.0
Cross-validation scores: [0.96666667 0.96666667 0.96666667 0.93333333 1.        ]
Mean of cross-validation score: 0.9666666666666666


I am getting a score of 1.0 and a Cross-validation scores of 0.967 for the SVM model which indicates that it has achieved perfect accuracy on the test set, correctly predicting the class labels for all test samples. This suggests that the model has effectively learned from the training data and can accurately classify new datapoints.

# Part 5: Neural Network

In [15]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp_model = MLPClassifier(hidden_layer_sizes=(500,), max_iter=1000, activation='tanh',random_state=42) 
mlp_model.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
sample_dataPoint = pd.DataFrame([[5, 3.4, 1.5, 0.2]], columns=["sepal_length", "sepal_width", "petal_length", "petal_width"])
probs = mlp_model.predict_proba(sample_dataPoint)
for class_name, prob in zip(mlp_model.classes_, probs[0]):
    print(class_name, ":", round(prob, 4))
# iii. Report on the score for the Neural Network, what does the score measure?
mlp_score = mlp_model.score(X_test, y_test)
print("mlp_score: ", mlp_score)
# iv: Experiment with different options for the neural network, report on your best configuration

param_grid = {
    'hidden_layer_sizes': [(800,), (500,), (100, 50), (200, 100, 50)],
    'activation': ['relu', 'logistic', 'tanh'],
    'max_iter':  [1000, 2000]
}
mlp = MLPClassifier(random_state=42)
grid_search = GridSearchCV(mlp, param_grid, cv=5)
grid_search.fit(X_train, y_train)

# best_params_ returns Parameter setting that gave the best results on the hold out data.
best_params = grid_search.best_params_
#best_score_ returns 
best_score = grid_search.best_score_

print("Best Parameters: ", best_params)
print("Best Cross-Validated Score: ", best_score)

# Testing the best esitmator which gave highest score on the data.
best_mlp_model = grid_search.best_estimator_
print("best mlp: ", best_mlp_model)
best_mlp_model.fit(X_train, y_train)

test_score = best_mlp_model.score(X_test, y_test)
print("Test Score with Best Parameters: ", test_score)

setosa : 0.998
versicolor : 0.002
virginica : 0.0
mlp_score:  0.8666666666666667
Best Parameters:  {'activation': 'logistic', 'hidden_layer_sizes': (800,), 'max_iter': 1000}
Best Cross-Validated Score:  0.9925925925925926
best mlp:  MLPClassifier(activation='logistic', hidden_layer_sizes=(800,), max_iter=1000,
              random_state=42)
Test Score with Best Parameters:  0.9333333333333333


The score reported by the neural network model (MLP Classifier) measures the accuracy of the model, which is the proportion of correct predictions made on the test data. It tells how well the model is performing by comparing its predicted labels with the actual labels. A score of 0.8667 (or 86.67%) means that the model correctly predicted about 87% of the test data points. This accuracy metric reflects how reliable the neural network is at identifying the correct class for new, unseen data based on what it learned from the training set.


Part iv Experimentation.

I used GridSearchCV to experiment with different options for neural network. I specified a param_grid with different hidden layers, activation functions, and max iterations. After performing the GridSearchCV the best parameters i got is {'activation': 'logistic', 'hidden_layer_sizes': (800,), 'max_iter': 1000} and it has a better score of 0.9334 than the score given previously which was 0.8667.

The logistic (sigmoid) function helped the model to better capture nonlinear relationships in the data as compared to relu and tanh.


A larger hidden layer with 800 units helped the model better capture complex patterns in the data. While larger layers can better recognize details, they also require more computing power.

Setting max_iter to 1000 allowed the model more time to converge fully. In previous runs, I observed that lower values led to convergence warnings, indicating that the model hadn’t reached its optimal state. The higher iteration count resolved this issue, allowing the model to stabilize and reach a more accurate solution.

# Part 6: K-Nearest Neighbors

In [ ]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
knn_model = KNeighborsClassifier(n_neighbors=5)
knn_model.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
sample_dataPoint = pd.DataFrame([[6.1,2.8,4.7,1.2]], columns=["sepal_length", "sepal_width", "petal_length", "petal_width"])
probs = knn_model.predict_proba(sample_dataPoint)
for class_name, prob in zip(knn_model.classes_, probs[0]):
    print(class_name, ":", round(prob, 4))
# iii. Report on the score for kNN, what does the score measure?
knn_score = knn_model.score(X_test, y_test)
print("knn_score: ",knn_score)

# generating the cross-validation score.
scores = cross_val_score(knn_model, X, y, cv=8)
print("cross_val_score",scores)
print("Mean of cross-validation score:", scores.mean())

setosa : 0.0
versicolor : 1.0
virginica : 0.0
knn_score:  0.8666666666666667
cross_val_score [1.         0.94736842 1.         0.89473684 0.94736842 0.94736842
 1.         1.        ]
Mean of cross-validation score: 0.9671052631578947


The score for the K-Nearest Neighbors (kNN) model measures the accuracy of the model, which is the percentage of correct predictions it made on the test data.I am getting a same score for the K-NN model as the one I got for Neural Network Model. 

A score of 0.8667 (or 86.67%) means the kNN model correctly predicted about 87% of the test data points. Since kNN relies on the similarity of nearby data points to make predictions, this score reflects how effective the model is at finding the correct class based on the closest neighbors in the feature space.

Also, The K-Nearest Neighbors (kNN) model achieved a cross-validation mean score of 0.967, indicating a high level of accuracy in classifying data points. This score shows that, on average, the model correctly predicted around 96.7% of cases across different portions of the dataset, demonstrating its strong performance on unseen data.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?


The score given by different models is as follows:

- Logistic Regression: 0.9334 
- Support Vector Machine: 1.0 and 0.967 (for cross validation)
- Neural Network: 0.8667 (for the tanh) and 0.9334 (for the logistic)
- K-Nearest Neighbors: 0.8667 


The model that performed the best on the dataset was the Support Vector Machine (SVM), achieving a perfect score of 1.0. This indicates that the SVM model was able to classify all the data points correctly, at least on this dataset.

Neural Network with the logistic activation function achieved the same score as logistic regression (0.9334), but with the tanh activation, its performance dropped to 0.8667. This could mean that the logistic function fits the data better, while tanh might be overfitting or underfitting certain areas.

Also, the difference in performance between neural network activation functions (e.g., tanh vs. logistic), along with variations in the number of hidden layers and maximum iterations, highlights how these parameters can significantly impact the results of a neural network.

K-Nearest Neighbors (KNN) also scored 0.8667. KNN often depends heavily on the neighborhood size, if I change the number of neighbors to 7 or 9 i am getting a better score.

Furthermore, The predicted probabilities for different classes are generally similar across the models, except for the K-Nearest Neighbors (kNN) model, which consistently assigns a probability of 1.0 to a single class and 0 to others. This behavior reflects kNN's tendency to make definitive classifications based on the closest neighbors.

One thing I find surprising is that the Support Vector Machine (SVM) model achieved a near-perfect score in a single evaluation but showed a slight drop in cross-validation. This contrasts with the other models, which generally performed better in cross-validation than in a single evaluation.



Also, The fact that logistic regression and the neural network with a logistic activation function have the same scores suggests that the dataset is easy to separate into classes with a simple line or boundary. This means we don’t need a complex model to get good results, as the patterns in the data are clear enough for a simpler model like logistic regression to perform just as well as the more complex neural network. 


This set of results emphasizes that simpler models might sometimes be as effective as complex ones for certain datasets.


